linalg/blockoperator.hpp

```diff
@@ -172,6 +172,96 @@ class BlockDiagonalPreconditioner : public Solver
+ * - Use SetBlock to fill the BlockOperator
+ * - Use the method Mult and MultTranspose to apply the operator to a vector.
+ *
+ * If a diagonal block is not set, it is assumed it is an identity block, if an off-diagonal block is not set, it is assumed to be a zero block.
+ *
+ */
+class BlockLowerTriangularPreconditioner : public Solver
+{
+public:
+   //! Constructor for BlockLowerTriangularPreconditioners with the same block-structure for rows and
+   //! columns.
+   /**
+    *  offsets: offsets that mark the start of each row/column block (size
+    *  nRowBlocks+1).  Note: BlockLowerTriangularPreconditioner will not own/copy the data contained
+    *  in offsets.
+    */
+   BlockLowerTriangularPreconditioner(const Array<int> & offsets);
+   //! Constructor for general BlockLowerTriangularPreconditioners.
+   /**
+    *  row_offsets: offsets that mark the start of each row block (size
+    *  nRowBlocks+1).  col_offsets: offsets that mark the start of each column
+    *  block (size nColBlocks+1).  Note: BlockLowerTriangularPreconditioner will not own/copy the
+    *  data contained in offsets.
+    */
+   BlockLowerTriangularPreconditioner(const Array<int> & row_offsets,
+                                      const Array<int> & col_offsets);
+
+   //! Add block op in the block-entry (iblock, iblock).
+   /**
+    * iblock: The block will be inserted in location (iblock, iblock).
+    * op: the Operator to be inserted.
+    */
+   void SetDiagonalBlock(int iblock, Operator *op);
+   //! Add a block op in the block-entry (iblock, jblock).
+   /**
+    * irow, icol: The block will be inserted in location (irow, icol).
+    * op: the Operator to be inserted.
+    */
+   void SetBlock(int iRow, int iCol, Operator *op);
+   //! This method is present since required by the abstract base class Solver
+   virtual void SetOperator(const Operator &op) { }
 
+   //! Return the number of blocks
+   int NumBlocks() const { return nRowBlocks; }
+
+   //! Return a reference to block i,j.
+   Operator & GetBlock(int iblock, int jblock)
+   { MFEM_VERIFY(op(iblock,jblock), ""); return *op(iblock,jblock); }
+
+   //! Return the offsets for block starts
+   Array<int> & Offsets() { return row_offsets; }
+
+   /// Operator application
+   virtual void Mult (const Vector & x, Vector & y) const;
+
+   /// Action of the transpose operator
+   virtual void MultTranspose (const Vector & x, Vector & y) const;
+
+   ~BlockLowerTriangularPreconditioner();
+
+   //! Controls the ownership of the blocks: if nonzero,
+   //! BlockLowerTriangularPreconditioner will delete all blocks that are set
+   //! (non-NULL); the default value is zero.
+   int owns_blocks;
+
+private:
+   //! Number of block rows
+   int nRowBlocks;
+   //! Number of block columns
+   int nColBlocks;
+   //! Row offsets for the starting position of each block
+   Array<int> row_offsets;
+   //! Column offsets for the starting position of each block
+   Array<int> col_offsets;
```

Comment:
I would suggest to remove 'col_offset', and rename 'row_offsets' simply as 'offsets' since the preconditioner makes sense only for square-block-operators with square diagonal blocks